File: PO/Project/C++/src/antelope.cpp
#include "antelope.h"
#include "world.h"
#include <cstdlib>

antelope::antelope(world* world_point, YX position)
:animal(4,4, position, world_point)
{
    avatar = 'A';
}

int antelope::Action() // 0 - normal move, 1 - collision detected
{
    int delta_y = rand() % 5 - 2; // Generates random number between -2 and 2
    int delta_x = rand() % 5 - 2; // Generates random number between -2 and 2

    if ((delta_x == 0) && (delta_y == 0)) {
        return 0;
    }

    return Move({delta_y, delta_x});
}

bool antelope::RejectAttak(organizm* attaker)
{
    if (rand() % 2 == 1) {
        world_point->GetLogger()->Log("antelope avoided attak");

        //find free field near it
        //if there is no free space, return false;
        field* new_lokation = world_point->GetFreeFieldNear(position);

        if (new_lokation == nullptr) {
            return  false;
        }

        world_point->FindField(position)->member = nullptr;

        attaker->GetWorld()->FindField(attaker->GetPosition())->member = nullptr;
        attaker->SetPosition(position);
        attaker->GetWorld()->FindField(attaker->GetPosition())->member = attaker;

        position = new_lokation->id;

        return true;
    }

    return false;
}

organizm* antelope::MakeChild(YX new_position)
{
    return (new antelope(world_point, new_position));
}

antelope::~antelope()
{
}

File: PO/Project/JavaProject/src/Animals/Antelope.java
package Animals;

import Abstractions.*;
import com.google.gson.JsonObject;

import java.awt.*;
import java.util.concurrent.ThreadLocalRandom;

public class Antelope extends Animal {

    public Antelope(World world, Point position) {
        super(4, 4, position, world);
        this.avatar = 'A';
        this.image_path = "Antelope.png";
        loadImage();
    }

    @Override
    public int action() {
        int deltaX, deltaY;
        do {
            deltaY = ThreadLocalRandom.current().nextInt(-2, 3); // Generates random number between -2 and 2
            deltaX = ThreadLocalRandom.current().nextInt(-2, 3); // Generates random number between -2 and 2
        } while ((deltaX == 0 && deltaY == 0) || (world.findField(new Point(position.x + deltaX, position.y + deltaY)) == null));

        if (deltaX == 0 && deltaY == 0) {
            return 0;
        }

        return move(new Point(deltaX, deltaY));
    }

    @Override
    public boolean rejectAttack(Organism attacker) {
        if (Math.random() < 0.5) {
            world.getLogger().log("Antelope avoided attack");

            Field newLocation = world.getFreeFieldNear(position);
            if (newLocation == null) {
                return false;
            }

            world.findField(position).setMember(null);

            attacker.getWorld().findField(attacker.getPosition()).setMember(null);
            attacker.setPosition(position);
            attacker.getWorld().findField(attacker.getPosition()).setMember(attacker);

            position = newLocation.getId();

            return true;
        }

        return false;
    }

    @Override
    public Organism makeChild(Point newPosition) {
        return new Antelope(world, newPosition);
    }
}

